src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -214,7 +215,40 @@ public TextWriter EndWritingScope()
                 _originalWriter = null;
             }
 
-            return writer;
+            var tagHelperContentWrapperTextWriter = new TagHelperContentWrapperTextWriter(Encoding.UTF8);
+            var razorWriter = writer as RazorTextWriter;
+            if (razorWriter != null)
+            {
+                razorWriter.CopyTo(tagHelperContentWrapperTextWriter);
+            }
+            else
+            {
+                tagHelperContentWrapperTextWriter.Write(writer.ToString());
```

Comment:
should also special-case the `StringCollectionTextWriter` created in `StartTagHelperWritingScope`.  the `ToString()` here loses many of the advantages of the Razor changes.